### PR TITLE
refactor: extract elevation routines to lib/shared/permissions.js

### PR DESCRIPTION
**Pull request by jviotti**
The elevation mechanism currently embedded in
`lib/child-writer/writer-proxy.js` is extracted as a separate re-usable
function in `lib/shared/permissions.js`.

This change hugely simplifies the writer proxy, while allowing us to
iterate faster on the elevation core functionality.

Signed-off-by: Juan Cruz Viotti <jviotti@openmailbox.org>

### CODE REVIEW

**Comment by jviotti** [comment]
OS testing:

- [x] macOS
- [ ] Windows
- [ ] GNU/Linux
- [ ] GNU/Linux on AppImage

File: lib/child-writer/writer-proxy.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016 resin.io
+ * Copyright 2017 resin.io
```
Comment (lurch) [inline lib/child-writer/writer-proxy.js]: By contrast, the files in this PR were created in 2016, so I wonder if their copyright-year should stay as 2016? I'm afraid I'm not a copyright lawyer, so I dunno what the "correct" solution is ;-)

EDIT: Did a quick search, and found http://stackoverflow.com/a/2391555 and https://www.gnu.org/prep/maintain/html_node/Copyright-Notices.html which at least seem to be consistent with each other.

File: tests/shared/permissions.spec.js [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+      afterEach(function() {
+        this.osPlatformStub.restore();
+      });
+
+      it('should return an empty array if no environment', function() {
+        m.chai.expect(permissions.getEnvironmentCommandPrefix()).to.deep.equal([]);
+      });
+
+      it('should return an empty array environment is an empty object', function() {
```
Comment (lurch) [inline tests/shared/permissions.spec.js]: Did your keyboard-gremlin eat an `if` from this description? ;)

File: lib/child-writer/writer-proxy.js [hunk 3]
```diff
@@ -135,44 +79,32 @@ return permissions.isElevated().then((elevated) => {
...
-
-    return sudoPrompt.execAsync(command, {
-      name: packageJSON.displayName
-    }).then((stdout, stderr) => {
-      if (!_.isEmpty(stderr)) {
-        throw errors.createError({
-          title: stderr
-        });
-      }
+    console.log(`Running: ${commandArguments.join(' ')}`);
```
Comment (lurch) [inline lib/child-writer/writer-proxy.js]: I wonder if this should be moved to `permissions.js` ?

File: tests/shared/permissions.spec.js [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
```
Comment (lurch) [inline tests/shared/permissions.spec.js]: if

File: tests/shared/permissions.spec.js [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+          '&&',
+          'call'
+        ]);
+      });
+
+      it('should ignore undefined and null variable values', function() {
+        m.chai.expect(permissions.getEnvironmentCommandPrefix({
+          FOO: null,
+          BAR: undefined,
+          BAZ: 'qux'
```
Comment (lurch) [inline tests/shared/permissions.spec.js]: Only a minor nit, but for unit-testing purposes perhaps you should rearrange the values as `null, 'qux', undefined` which checks that valid-values work after null, and also checks that undefined values are ignored _after_ a valid value?

File: tests/shared/permissions.spec.js [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+      afterEach(function() {
+        this.osPlatformStub.restore();
+      });
+
+      it('should return an empty array if no environment', function() {
+        m.chai.expect(permissions.getEnvironmentCommandPrefix()).to.deep.equal([]);
+      });
+
+      it('should return an empty array environment is an empty object', function() {
```
Comment (lurch) [inline tests/shared/permissions.spec.js]: Pesky gremlin!

**Comment by lurch** [comment]
On Windows, did you test both portable and non-portable versions?
For all platforms, did you check the behaviour when Etcher is already launched with sudo/admin rights? (which it seems that some users still do, even though it's not necessary)

**Comment by jviotti** [comment]
Yep, it all seems to work fine :)

**Comment by jviotti** [comment]
Oh damn, I merged, but didn't notice all the comments that where there! I'll send another PR with the fixes!

File: lib/child-writer/writer-proxy.js (hunk 1 above)
Comment (jviotti) [inline lib/child-writer/writer-proxy.js]: Oh, great find :+1:

File: lib/child-writer/writer-proxy.js (hunk 3 above)
Comment (jviotti) [inline lib/child-writer/writer-proxy.js]: Can we do it once we get analytics support outside the main Electron process, given that will take care of not printing the lines during tests, etc?

File: tests/shared/permissions.spec.js (hunk 2 above)
Comment (jviotti) [inline tests/shared/permissions.spec.js]: Haha, fixed!

File: tests/shared/permissions.spec.js (hunk 5 above)
Comment (jviotti) [inline tests/shared/permissions.spec.js]: Fixed!

File: tests/shared/permissions.spec.js (hunk 4 above)
Comment (jviotti) [inline tests/shared/permissions.spec.js]: Fixed!

File: tests/shared/permissions.spec.js (hunk 6 above)
Comment (jviotti) [inline tests/shared/permissions.spec.js]: Haha, fixed!

**Comment by jviotti** [comment]
See https://github.com/resin-io/etcher/pull/1355. Sorry about that; the approval confused me!

File: lib/child-writer/writer-proxy.js (hunk 1 above)
Comment (lurch) [inline lib/child-writer/writer-proxy.js]: Perhaps something to add to the Contributing guide? And maybe the contributing guide should also say that by submitting a PR to the Etcher repo, any 3rd-party authors are explicitly assigning their copyright over to resin.io?

File: lib/child-writer/writer-proxy.js (hunk 3 above)
Comment (lurch) [inline lib/child-writer/writer-proxy.js]: Entirely up to you, it was only a suggestion ;-)

**Comment by lurch** [comment]
Okay, do I need to be less liberal with my approvals in future? ;)

**Comment by jviotti** [comment]
Haha, don't worry, it was totally my fault.

**Pull request merged**